modules/join/src/main/java/org/apache/ignite/internal/join/ClusterManagementGroupManager.java

```diff
@@ -0,0 +1,198 @@
+
+                    raftManager.prepareRaftGroup(CMG_RAFT_GROUP_NAME, nodes, CmgRaftGroupListener::new)
+                            .whenComplete((service, e) -> {
+                                if (e == null) {
+                                    Peer leader = service.leader();
+
+                                    assert leader != null;
+
+                                    messagingService.respond(addr, leaderElectedResponse(msgFactory, leader), correlationId);
+                                } else {
+                                    messagingService.respond(addr, errorResponse(msgFactory, e), correlationId);
+                                }
+                            });
+                }
+            } catch (Exception e) {
+                log.error("Exception when initializing the CMG", e);
+
+                if (correlationId != null) {
+                    messagingService.respond(addr, errorResponse(msgFactory, e), correlationId);
+                }
+            } finally {
+                busyLock.leaveBusy();
+            }
+        });
+    }
+
+    /**
+     * Initializes the cluster that this node is present in.
+     *
+     * @param metaStorageNodeNames names of nodes that will host the Meta Storage.
+     * @param cmgNodeNames names of nodes that will host the Cluster Management Group.
+     */
+    public Leaders initCluster(Collection<String> metaStorageNodeNames, Collection<String> cmgNodeNames) throws NodeStoppingException {
+        if (!busyLock.enterBusy()) {
+            throw new NodeStoppingException();
+        }
+
+        try {
+            return clusterInitializer.initCluster(metaStorageNodeNames, cmgNodeNames).get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+
+            throw new IgniteInternalException("Interrupted while initializing the cluster", e);
+        } catch (ExecutionException e) {
+            throw new IgniteInternalException("Unable to initialize the cluster", e.getCause());
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    private static NetworkMessage errorResponse(InitMessagesFactory msgFactory, Throwable e) {
+        log.error("Exception when starting the CMG", e);
+
+        return msgFactory.initErrorMessage()
+                .cause(e.getMessage())
+                .build();
+    }
+
+    private NetworkMessage leaderElectedResponse(InitMessagesFactory msgFactory, Peer leader) {
+        ClusterNode leaderNode = clusterService.topologyService().getByAddress(leader.address());
+
+        assert leaderNode != null;
+
+        log.info("CMG leader elected: " + leaderNode.name());
+
+        return msgFactory.leaderElectedMessage()
+                .leaderName(leaderNode.name())
+                .build();
+    }
+
+    @Override
+    public void start() {
+        restModule.addRoute(new Route(
```

Comment:
Why is init command handler initialized in `ClusterManagementGroupManager`? It seems a bit weird. Would a separate service designed specifically for init handling be better?

Reply:
> It seems a bit weird

I think it's fine since this manager is responsible for the init procedure. We can introduce a separate manager for this purposes, but I think it will just bloat the overall amount of components unnecessarily. What do you think?

Reply:
The thing that bothers me here is that, judging by the class name, it should be responsible only for a part of the init process (namely, the CMG init), but in reality it also takes part in other init activities which are not part of the CMG init. For someone who will (in the future) search for this REST handler it will be very surprising to finally find it here.
If you don't like the idea of splitting the class, how about renaming it so that it's not CMG-specific anymore?